BUG: avoid double superscipts with unicode

MridulS opened the pull request:
Fixes https://github.com/astropy/astropy/issues/14403

I have added a check for only a subset of units (the ones being tested in latex unit tests). I'm not sure if this is an exhaustive list :)

pllim commented on astropy/units/format/unicode_format.py:
```diff
@@ -5,7 +5,9 @@
 
 
-from . import console
+import astropy.units as u
```
I think this is preferred?
```suggestion
from astropy import units as u
```

mhvk replied:
I don't think we need the import (see other comment) but we seem to have different "defaults" for this. I seem to have settled on using `from ... import X` for cases where `X` is not a module...

pllim commented on astropy/units/format/unicode_format.py:
```diff
@@ -5,7 +5,9 @@
-from . import console
+import astropy.units as u
+
+from . import console, utils
```
Is there any benefit to use local import over absolute import nowadays?

mhvk replied:
I think following precedent in `latex` is fine. I personally like that it is immediately clear that it is a related module (but am very happy we got rid of using `..`!)

MridulS replied:
There are relative imports in a bunch of places, and I didn't want to touch any linting choices here 😅 
Does it make sense to send a new PR which fixes relative imports all across astropy?

mhvk replied:
I think you made the right choice, certainly for this PR. For changing imports across astropy, I think there are better things to do with both your and reviewer time - I think we are in a decent state and to be honest I'm getting a bit tired of the churn caused by trying to standardize on what sometimes seems to be just the latest fashion. More practically, at some point `units` should really be split of into a standalone package, and then having those relative imports means less work...

pllim commented on astropy/units/tests/test_format.py:
I find that parametrization makes the test harder to read. Is this necessary?

pllim commented on docs/changes/units/15755.bugfix.rst:
Might want to rephrase a little bit so it is easier for end-users to understand. Not all users are familiar with the internals or jargons like "repr".

pllim reviewed (commented):
Thanks! Just a general review.

mhvk commented on astropy/units/format/unicode_format.py:
```diff
@@ -33,6 +35,23 @@ class Unicode(console.Console):
+            u.electron,
+        ]
+        name = cls._get_unit_name(unit)
+        if power != 1 and unit in superscript_unicode_units:
```
I'd follow the latex one a bit more closely, like:
```
name = cls._get_unit_name(unit)
if power != 1:
    if (short_name := unit.short_names[0]) in ("arcsec", "arcmin", "deg", "hourangle", "electron"):
        name  = short_name
    name += cls._format_superscript(utils.format_power(power))
return name
```
This has the advantage of not having to import `units` and making the comparison much faster (unit equality, which is implicit in your `unit in superscript_unicode_units` is much slower than string comparison).

p.s. Actually, it would be fine as well to just `if name == (..., "e⁻")`, i.e., compare with the unicode units we want to avoid...

MridulS replied:
Yup, that makes sense.

mhvk commented on astropy/units/tests/test_format.py:
```diff
@@ -815,18 +815,53 @@ def test_fits_scale_factor_errors():
 
 
-def test_double_superscript():
+@pytest.mark.parametrize(
```
I agree with @pllim that this does not make things clearer. How about parametrizing as.
```
"unit, latex, unicode",
[
    (u.deg, r"$\mathrm{{}^{\circ}}$", "°"),
    (u.deg**2, ...)
]
```
and then obviously,
```
assert unit.to_string("latex") == latex
assert unit.to_string("unicode") == unicode
```

That way we keep clear what expected string belongs to what unit.

mhvk reviewed (commented):
Thanks, @MridulS. Basics looks good, but some suggestions in-line. Agreed with @pllim that the changelog entry can be written a bit more with end users in mind.

MridulS commented:
Thanks for the reviews @pllim @mhvk, I have addressed the comments in https://github.com/astropy/astropy/pull/15755/commits/659576b8dcaf23a272cfb4c4fc66fbfb661e8256

mhvk commented on astropy/units/format/unicode_format.py:
```diff
@@ -33,6 +31,17 @@ class Unicode(console.Console):
+        # Check for superscript units
+        if power != 1 and name in ("°", "e⁻", "″", "′", "ʰ"):
+            name = unit.short_names[0]
+            name += cls._format_superscript(utils.format_power(power))
```
Generally, I like to punt to `super()` when possible, but here it feels a bit superfluous since it is not done for all cases and, as is, means that one has to get `name` twice. So, I think it is better to follow `latex` and make it more specific, as I suggested in my earlier comment, which, with your change to `name in ...` (better!) is,
```
name = cls._get_unit_name(unit)
if power != 1:
    if name in ("°", "e⁻", "″", "′", "ʰ"):
        name  = short_name
    name += cls._format_superscript(utils.format_power(power))
return name
```

mhvk commented on docs/changes/units/15755.bugfix.rst:
```diff
@@ -0,0 +1 @@
+Fix issue with unicode string representations of units (with superscripts) when printing them in combination with power. The unit now behaves consistently with LaTeX representation and uses the literal short name instead of superscript unicode character in combination with power.
```
Since my other comment asks for a change, could you reformat this to a maximum line length of 80? Maybe slightly rephrase too:
```
Fix an issue with unicode string representations of units shown as
superscripts (like degree) when raised to some power. Like for
LaTeX representations, now the superscript unicode character is
replaced by the literal short name before adding the power.
```

mhvk reviewed (commented):
Thanks! Two smaller comments left... (probably getting carried away a bit, sorry!)

mhvk reviewed (approved):
Super, thanks! All good now.
